akvo/rsr/management/commands/report_d2d.py

```diff
@@ -14,89 +14,61 @@
 
 from akvo.rsr.models import (
     IatiExport, Organisation, Project, IndicatorPeriodData,
-    IndicatorPeriodDataComment, LoginLog, PublishingStatus,
-    ProjectUpdate
+    LoginLog, PublishingStatus,
 )
 
 
 def get_active_organisations(date_start, date_end):
-    has_project_updates = set(
-        ProjectUpdate.objects
-        .filter(Q(created_at__range=(date_start, date_end)) | Q(last_modified_at__range=(date_start, date_end)))
-        .values_list('project__id', flat=True)
-    )
-    has_indicator_updates = set(
-        IndicatorPeriodData.objects
-        .filter(Q(created_at__range=(date_start, date_end)) | Q(last_modified_at__range=(date_start, date_end)))
-        .values_list('period__indicator__result__project__id', flat=True)
-    )
-    pids = has_project_updates | has_indicator_updates
-    return set(Project.objects.filter(id__in=pids).values_list('primary_organisation__id', flat=True))
+    projects_with_project_updates = Project.objects.filter(
+        Q(project_updates__created_at__range=(date_start, date_end))
+        | Q(project_updates__last_modified_at__range=(date_start, date_end))
+    ).distinct()
+    projects_with_indicator_updates = Project.objects.filter(
+        Q(results__indicators__periods__data__created_at__range=(date_start, date_end))
+        | Q(results__indicators__periods__data__last_modified_at__range=(date_start, date_end))
+    ).distinct()
+    active_projects = projects_with_project_updates.union(projects_with_indicator_updates)
+    org_ids_of_active_projects = active_projects.values_list('primary_organisation', flat=True)
+    return Organisation.objects.filter(id__in=set(org_ids_of_active_projects))
```

Comment:
It should be possible to do `Organisation.objects.filter(projects__in=active_projects)`. `values_list` isn't necessary with [`in` lookups](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#in)

Reply:
```diff
diff --git a/akvo/rsr/management/commands/report_d2d.py b/akvo/rsr/management/commands/report_d2d.py
index 591f8cad3..7f8174267 100644
--- a/akvo/rsr/management/commands/report_d2d.py
+++ b/akvo/rsr/management/commands/report_d2d.py
@@ -28,8 +28,7 @@ def get_active_organisations(date_start, date_end):
         | Q(results__indicators__periods__data__last_modified_at__range=(date_start, date_end))
     ).distinct()
     active_projects = projects_with_project_updates.union(projects_with_indicator_updates)
-    org_ids_of_active_projects = active_projects.values_list('primary_organisation', flat=True)
-    return Organisation.objects.filter(id__in=set(org_ids_of_active_projects))
+    return Organisation.objects.filter(id__in=active_projects)
```

And I got this error

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.SyntaxError: subquery has too many columns
LINE 1: ... "rsr_organisation" WHERE "rsr_organisation"."id" IN ((SELEC...
                                                             ^


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "./manage.py", line 27, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.8/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.8/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/var/akvo/rsr/code/akvo/rsr/management/commands/report_d2d.py", line 138, in handle
    print(f'# of active partners using RSR: {active_partners.count()}')
  File "/usr/local/lib/python3.8/site-packages/django/db/models/query.py", line 412, in count
    return self.query.get_count(using=self.db)
  File "/usr/local/lib/python3.8/site-packages/django/db/models/sql/query.py", line 519, in get_count
    number = obj.get_aggregation(using, ['__count'])['__count']
  File "/usr/local/lib/python3.8/site-packages/django/db/models/sql/query.py", line 504, in get_aggregation
    result = compiler.execute_sql(SINGLE)
  File "/usr/local/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1175, in execute_sql
    cursor.execute(sql, params)
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 98, in execute
    return super().execute(sql, params)
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.8/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: subquery has too many columns
LINE 1: ... "rsr_organisation" WHERE "rsr_organisation"."id" IN ((SELEC...
 ```

where did it go wrong?

Reply:
`Organisation.objects.filter(projects__in=active_projects)` isn't what you want?

Reply:
```diff
diff --git a/akvo/rsr/management/commands/report_d2d.py b/akvo/rsr/management/commands/report_d2d.py
index 591f8cad3..0a2eeb8c4 100644
--- a/akvo/rsr/management/commands/report_d2d.py
+++ b/akvo/rsr/management/commands/report_d2d.py
@@ -28,8 +28,7 @@ def get_active_organisations(date_start, date_end):
         | Q(results__indicators__periods__data__last_modified_at__range=(date_start, date_end))
     ).distinct()
     active_projects = projects_with_project_updates.union(projects_with_indicator_updates)
-    org_ids_of_active_projects = active_projects.values_list('primary_organisation', flat=True)
-    return Organisation.objects.filter(id__in=set(org_ids_of_active_projects))
+    return Organisation.objects.filter(projects__in=active_projects)

```

Got this error

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.SyntaxError: subquery has too many columns
LINE 1: ...isation_id") WHERE "rsr_partnership"."project_id" IN ((SELEC...
                                                             ^


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "./manage.py", line 27, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.8/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.8/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/var/akvo/rsr/code/akvo/rsr/management/commands/report_d2d.py", line 136, in handle
    print(f'# of active partners using RSR: {active_partners.count()}')
  File "/usr/local/lib/python3.8/site-packages/django/db/models/query.py", line 412, in count
    return self.query.get_count(using=self.db)
  File "/usr/local/lib/python3.8/site-packages/django/db/models/sql/query.py", line 519, in get_count
    number = obj.get_aggregation(using, ['__count'])['__count']
  File "/usr/local/lib/python3.8/site-packages/django/db/models/sql/query.py", line 504, in get_aggregation
    result = compiler.execute_sql(SINGLE)
  File "/usr/local/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1175, in execute_sql
    cursor.execute(sql, params)
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 98, in execute
    return super().execute(sql, params)
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.8/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: subquery has too many columns
LINE 1: ...isation_id") WHERE "rsr_partnership"."project_id" IN ((SELEC...
                                                             ^
```

Where did it go wrong?

Reply:
And the [`primary_organisation`](https://github.com/akvo/akvo-rsr/blob/master/akvo/rsr/models/project.py#L315-L316) field don't have the `related_name` attribute. It also has a different purpose than the [`partners`](https://github.com/akvo/akvo-rsr/blob/master/akvo/rsr/models/project.py#L170-L173)

Reply:
As discuss, it seems more complicated than what I suggested. We'll keep this as it is.